docs/design_doc.md

```diff
@@ -0,0 +1,300 @@
+
+## Robot Components
+
+<center><img src="framework.png" width=900></center>
+A **Robot** consists of some **computation tasks**, i.e., applying some **algorithms** to using or learning **models**. For example, a robot has one reinforcement learning task and it uses an Actor-Critic algorithm to learn the policy model. To provide additional training of the policy model, the robot can also have some auxiliary task which is usually dealt with some unsupervised learning algorithm. Another scenario is where a robot needs different policies at different levels of temporal abstraction. In this case, we can assign one policy learning task to each policy. 
+
+In the following, we describe the design details of  the above-mentioned major components of robot. The the design principle is that we want to provide users with a framework that is easy for them to implement their own robots, by overriding or extending some of these components. The components of robot can be divided into two groups. One group handles the actual models, e.g., model computation, parameter sharing and synchronization; we call it **computation group**. The other group interacts with environment and produce/consume data to/from *computation group*; we call it **agent group** (we will explain this name later).
+
+Let's first talk about the components in **computation group**: `ComputationTask`, `Algorithm` and `Model`. 
+
+### Computation Group
+
+A robot consists of multiple `ComputationTask`s (CTs) executed in some order. Each CT can have its own data I/O interaction with outside. One CT may have I/O interaction with the other CTs. Intuitively, a ComputationTask should perform a relatively independent computation routine. For example, in hierarchical RL, one CT predicts the goal and the next CT predicts the actions according to the goal. We require that 
+
+- The code between two CTs is non-differentiable computation or control.
+- There is no gradient flowing between two adjacent CTs.
+
+If you have two modules that must have gradients flow between them, then consider puting the two modules in one CT.
+Finally, different CTs might have different time resolutions. For example, a CT outputs actions at every time step, which another CT outputs actions very 100 time steps.
+
+<center><img src="ct.png"><img src="model.png"></center>
+
+<center>[**In the above figure, blue boxes represent pure Python logic control and computation, and purple boxes represent logic control and computation in Fluid programs.**](#)</center>
+
+A `ComputationTask` consists of two `Algorithm`s, a target algorithm and a reference algorithm. An algorithm should implement general prediction and learning functions, independent of the actual problem. The core of an algorithm is usually the back-end (optimization) of a network. An algorithm further contains a model which is problem-dependent and allows the user to write the front-end (data perception) of the network. 
```

Comment:
No need to have concepts of "target algorithm" and "reference algorithm". Target and/or reference should be details handled in algorithm, not in computation task